service/s3/presign_post.go

```diff
@@ -0,0 +1,408 @@
+package s3
+
+import (
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	awsmiddleware "github.com/aws/aws-sdk-go-v2/aws/middleware"
+	"github.com/aws/aws-sdk-go-v2/aws/retry"
+	v4 "github.com/aws/aws-sdk-go-v2/aws/signer/v4"
+	internalcontext "github.com/aws/aws-sdk-go-v2/internal/context"
+	"github.com/aws/aws-sdk-go-v2/internal/sdk"
+	acceptencodingcust "github.com/aws/aws-sdk-go-v2/service/internal/accept-encoding"
+	presignedurlcust "github.com/aws/aws-sdk-go-v2/service/internal/presigned-url"
+	"github.com/aws/smithy-go/middleware"
+	smithyhttp "github.com/aws/smithy-go/transport/http"
+)
+
+const (
+	algorithmHeader  = "X-Amz-Algorithm"
+	credentialHeader = "X-Amz-Credential"
+	dateHeader       = "X-Amz-Date"
+	tokenHeader      = "X-Amz-Security-Token"
+	signatureHeader  = "X-Amz-Signature"
+
+	algorithm        = "AWS4-HMAC-SHA256"
+	aws4Request      = "aws4_request"
+	bucketHeader     = "bucket"
+	defaultExpiresIn = 15 * time.Minute
+)
+
+// PresignedPostRequest represents a presigned request to be sent using HTTP verb POST and FormData
+type PresignedPostRequest struct {
+	// Represents the Base URL to make a request to
+	URL string
+	// Values is a key-value map of values to be sent as FormData
+	// these values are not encoded
+	Values map[string]interface{}
+}
+
+// postSignAdapter adapter to implement the presignPost interface
+type postSignAdapter struct{}
+
+// PresignPost creates a special kind of [presigned request]
+// to be used with HTTP verb POST.
+// It differs from PUT request mostly on
+// 1. It accepts a new set of parameters, `Conditions[]`, that are used to create a policy doc to limit where an object can be posted to
+// 2. The return value needs to have more processing since it's meant to be sent via a form and not stand on its own
+// 3. There's no body to be signed, since that will be attached when the actual request is made
+// 4. The signature is made based on the policy document, not the whole request
+// More information can be found at https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPOST.html
+//
+// [presigned request] https://docs.aws.amazon.com/AmazonS3/latest/userguide/using-presigned-url.html
+func (s *postSignAdapter) PresignPost(
+	credentials aws.Credentials,
+	bucket string, key string,
+	region string, service string, signingTime time.Time, conditions []interface{}, expirationTime time.Time, optFns ...func(*v4.SignerOptions),
+) (fields map[string]interface{}, err error) {
+	if conditions == nil {
```

Comment:
This isn't necessary, `nil` slices and 0-length slices are programmatically treated basically the same everywhere (including in an `append(x, y...)` call where you're using it)

Reply:
Removed